### ISSUE TITLE
error: unknown type name 'offs_t' during 0.259 build on FreeBSD

### ISSUE THREAD

**Issue by github-c999-bfa4-cc81** [mamedev/mame#11591]
### MAME version

0.259

### System information

FreeBSD-12.4, 13.2 releases, and 15-CURRENT on amd64, and i386.

### Incorrect behaviour

Clang-15, and 16 both fail asserting that the types offs_t, and u32 are undeclared in [src/devices/cpu/drcfe.h](https://github.com/mamedev/mame/blob/0d9eaa46ffb578cc4a5b884a4dfab5e14b3c9d5f/src/devices/cpu/drcfe.h):

`Script started, output file is /tmp/mame.log
===>  Building for mame-0.259
gmake[2]: Entering directory '/tmp/mame-0.256/net/fafnir/git/freebsd/ports/main/emulators/mame/work/mame-mame0259'
Clang 15.0.7 detected
gmake -C build/projects/sdl/mame/gmake-freebsd-clang config=release64 precompile
gmake[3]: Entering directory '/tmp/mame-0.256/net/fafnir/git/freebsd/ports/main/emulators/mame/work/mame-mame0259/build/projects/sdl/mame/gmake-freebsd-clang'
gmake[3]: Leaving directory '/tmp/mame-0.256/net/fafnir/git/freebsd/ports/main/emulators/mame/work/mame-mame0259/build/projects/sdl/mame/gmake-freebsd-clang'
gmake -C build/projects/sdl/mame/gmake-freebsd-clang config=release64
gmake[3]: Entering directory '/tmp/mame-0.256/net/fafnir/git/freebsd/ports/main/emulators/mame/work/mame-mame0259/build/projects/sdl/mame/gmake-freebsd-clang'
Compiling src/devices/cpu/sh/sh7021.cpp...
In file included from ../../../../../src/devices/cpu/sh/sh7021.cpp:4:
In file included from ../../../../../src/devices/cpu/sh/sh7021.h:9:
In file included from ../../../../../src/devices/cpu/sh/sh2.h:20:
In file included from ../../../../../src/devices/cpu/sh/sh.h:9:
../../../../../src/devices/cpu/drcfe.h:43:11: error: unknown type name 'offs_t'
constexpr offs_t BRANCH_TARGET_DYNAMIC = ~offs_t(0);
          ^
../../../../../src/devices/cpu/drcfe.h:43:43: error: use of undeclared identifier 'offs_t'
constexpr offs_t BRANCH_TARGET_DYNAMIC = ~offs_t(0);
                                          ^
../../../../../src/devices/cpu/drcfe.h:47:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_IS_UNCONDITIONAL_BRANCH = 0x00000001;       // instruction is unconditional branch
          ^
../../../../../src/devices/cpu/drcfe.h:48:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_IS_CONDITIONAL_BRANCH   = 0x00000002;       // instruction is conditional branch
          ^
../../../../../src/devices/cpu/drcfe.h:49:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_IS_BRANCH               = (OPFLAG_IS_UNCONDITIONAL_BRANCH | OPFLAG_IS_CONDITIONAL_BRANCH);
          ^
../../../../../src/devices/cpu/drcfe.h:50:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_IS_BRANCH_TARGET        = 0x00000004;       // instruction is the target of a branch
          ^
../../../../../src/devices/cpu/drcfe.h:51:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_IN_DELAY_SLOT           = 0x00000008;       // instruction is in the delay slot of a branch
          ^
../../../../../src/devices/cpu/drcfe.h:52:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_INTRABLOCK_BRANCH       = 0x00000010;       // instruction branches within the block
          ^
../../../../../src/devices/cpu/drcfe.h:55:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_CAN_TRIGGER_SW_INT      = 0x00000020;       // instruction can trigger a software interrupt
          ^
../../../../../src/devices/cpu/drcfe.h:56:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_CAN_EXPOSE_EXTERNAL_INT = 0x00000040;       // instruction can expose an external interrupt
          ^
../../../../../src/devices/cpu/drcfe.h:57:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_CAN_CAUSE_EXCEPTION     = 0x00000080;       // instruction may generate exception
          ^
../../../../../src/devices/cpu/drcfe.h:58:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_WILL_CAUSE_EXCEPTION    = 0x00000100;       // instruction will generate exception
          ^
../../../../../src/devices/cpu/drcfe.h:59:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_PRIVILEGED              = 0x00000200;       // instruction is privileged
          ^
../../../../../src/devices/cpu/drcfe.h:62:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_VALIDATE_TLB            = 0x00000400;       // instruction must validate TLB before execution
          ^
../../../../../src/devices/cpu/drcfe.h:63:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_MODIFIES_TRANSLATION    = 0x00000800;       // instruction modifies the TLB
          ^
../../../../../src/devices/cpu/drcfe.h:64:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_COMPILER_PAGE_FAULT     = 0x00001000;       // compiler hit a page fault when parsing
          ^
../../../../../src/devices/cpu/drcfe.h:65:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_COMPILER_UNMAPPED       = 0x00002000;       // compiler hit unmapped memory when parsing
          ^
../../../../../src/devices/cpu/drcfe.h:68:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_INVALID_OPCODE          = 0x00004000;       // instruction is invalid
          ^
../../../../../src/devices/cpu/drcfe.h:69:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_VIRTUAL_NOOP            = 0x00008000;       // instruction is a virtual no-op
          ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
gmake[4]: *** [optional.make:26496: ../../../../freebsd/obj/x64/Release/src/devices/cpu/sh/sh7021.o] Error 1
gmake[3]: *** [Makefile:79: optional] Error 2
gmake[3]: Leaving directory '/tmp/mame-0.256/net/fafnir/git/freebsd/ports/main/emulators/mame/work/mame-mame0259/build/projects/sdl/mame/gmake-freebsd-clang'
gmake[2]: *** [makefile:1421: freebsd_x64_clang] Error 2
gmake[2]: Leaving directory '/tmp/mame-0.256/net/fafnir/git/freebsd/ports/main/emulators/mame/work/mame-mame0259'
*** Error code 1

Stop.
make[1]: stopped in /net/fafnir/git/freebsd/ports/main/emulators/mame
*** Error code 1

Stop.
make: stopped in /net/fafnir/git/freebsd/ports/main/emulators/mame

Script done, output file is /tmp/mame.log`

Considering MAME's official support for platforms does not include FreeBSD, I am unsure if this in incorrect behaviour.

### Expected behaviour

Suspecting this is a FreeBSD only issue, seeing as OpenBSD does not have a similar patch, and they also use LLVM to build MAME, I am unsure what to do, other than try here. The problem was not present prior to 0.259 on any of FreeBSD supported releases of the MAME Port. The expectation, given a standards compliant C++ program, the program should only require a recompile on systems also supporting the same C++ toolchain. 

'src/devices/cpu/drcfe.h' needs visibility of the types offs_t and u32, which I believe (thanks to some help via IRC) comes from [src/emu/emumem.h](https://github.com/mamedev/mame/blob/0d9eaa46ffb578cc4a5b884a4dfab5e14b3c9d5f/src/emu/emumem.h), and it looks like 'src/devices/cpu/sh/sh.h', has visibility, so not sure what to do. I added 'emu.h' to 'src/devices/cpu/sh/sh.h' believing this is the root of that build pass, and compilation completes without error.

### Additional details

Current local patch:
`
--- src/devices/cpu/sh/sh.h.orig        2023-10-03 01:49:35 UTC
+++ src/devices/cpu/sh/sh.h
@@ -6,6 +6,7 @@

 #pragma once

+#include "emu.h"
 #include "cpu/drcfe.h"
 #include "cpu/drcuml.h"
`

Thanks.

**Comment by cuavas** [comment]
Fixed by 4a7950dc1f0f6007b164a34f3fb3a83b2b88b79f.  You must be building with precompiled prefix headers disabled while the OpenBSD people have them enabled.

**Comment by github-c999-bfa4-cc81** [comment]
Thank you, the build continues with the fix backported to 0.259, however it fails further on:

```
Compiling src/devices/cpu/sh/sh7014.cpp...                                                                                                                                                                                                           [22/16009]
Compiling src/devices/cpu/sh/sh7014_bsc.cpp...
Compiling src/devices/cpu/sh/sh7014_dmac.cpp...
Compiling src/devices/cpu/sh/sh7014_intc.cpp...
Compiling src/devices/cpu/sh/sh7014_mtu.cpp...
Compiling src/devices/cpu/sh/sh7014_port.cpp...
Compiling src/devices/cpu/sh/sh7014_sci.cpp...
Compiling src/devices/cpu/sh/sh7021.cpp...
Compiling src/devices/cpu/sh/sh7032.cpp...
Compiling src/devices/cpu/sh/sh7604.cpp...
Compiling src/devices/cpu/sh/sh7604_bus.cpp...
Compiling src/devices/cpu/sh/sh7604_sci.cpp...
Compiling src/devices/cpu/sh/sh7604_wdt.cpp...
Compiling src/devices/cpu/sh/sh_fe.cpp...
In file included from ../../../../../src/devices/cpu/sh/sh7032.cpp:4:
In file included from ../../../../../src/devices/cpu/sh/sh7032.h:9:
In file included from ../../../../../src/devices/cpu/sh/sh2.h:20:
In file included from ../../../../../src/devices/cpu/sh/sh.h:9:
../../../../../src/devices/cpu/drcfe.h:42:11: error: unknown type name 'offs_t'
constexpr offs_t BRANCH_TARGET_DYNAMIC = ~offs_t(0);
          ^
../../../../../src/devices/cpu/drcfe.h:42:43: error: use of undeclared identifier 'offs_t'
constexpr offs_t BRANCH_TARGET_DYNAMIC = ~offs_t(0);
                                          ^
../../../../../src/devices/cpu/drcfe.h:46:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_IS_UNCONDITIONAL_BRANCH = 0x00000001;       // instruction is unconditional branch
          ^
../../../../../src/devices/cpu/drcfe.h:47:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_IS_CONDITIONAL_BRANCH   = 0x00000002;       // instruction is conditional branch
          ^
../../../../../src/devices/cpu/drcfe.h:48:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_IS_BRANCH               = (OPFLAG_IS_UNCONDITIONAL_BRANCH | OPFLAG_IS_CONDITIONAL_BRANCH);
          ^
../../../../../src/devices/cpu/drcfe.h:49:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_IS_BRANCH_TARGET        = 0x00000004;       // instruction is the target of a branch
          ^
../../../../../src/devices/cpu/drcfe.h:50:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_IN_DELAY_SLOT           = 0x00000008;       // instruction is in the delay slot of a branch
          ^
../../../../../src/devices/cpu/drcfe.h:51:11: error: unknown type name 'u32'
constexpr u32 OPFLAG_INTRABLOCK_BRANCH       = 0x00000010;       // instruction branches within the block
```

Would src/devices/cpu/sh/sh7032.cpp be the next place to add emu.h?

**Comment by github-c999-bfa4-cc81** [comment]
Sorry for the noise, did not backport entire patch.
